From 982627fdba70aec3a6dc2af7dae184ce275783f1 Mon Sep 17 00:00:00 2001
From: Uoti Urpala <uau@mplayer2.org>
Date: Wed, 3 Oct 2012 03:18:16 +0300
Subject: [PATCH 8/9] mixer: fix crash when muted and audio init fails

uninit_player() called both mixer_uninit() and ao_uninit() under
"if (mpctx->ao)". However, if AO or filter chain initialization fails
in reinit_audio_chain(), then mpctx->ao is set but mpctx->mixer->ao is
not. mixer_uninit() crashed in this case. Make mixer_uninit() a no-op
if no AO is set in the mixer. Also make the call to mixer_unit() in
uninit_player() unconditional, as calling it is now always safe.
---
 mixer.c   | 3 +++
 mplayer.c | 5 ++---
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/mixer.c b/mixer.c
index 7be2179..9749e0b 100644
--- a/mixer.c
+++ b/mixer.c
@@ -265,6 +265,9 @@ void mixer_reinit(struct mixer *mixer, struct ao *ao)
  */
 void mixer_uninit(struct mixer *mixer)
 {
+    if (!mixer->ao)
+        return;
+
     checkvolume(mixer);
     if (mixer->muted_by_us) {
         /* Current audio output API combines playing the remaining buffered
diff --git a/mplayer.c b/mplayer.c
index 3076d78..7e0ca02 100644
--- a/mplayer.c
+++ b/mplayer.c
@@ -671,10 +671,9 @@ void uninit_player(struct MPContext *mpctx, unsigned int mask)
     if (mask & INITIALIZED_AO) {
         mpctx->initialized_flags &= ~INITIALIZED_AO;
         current_module = "uninit_ao";
-        if (mpctx->ao) {
-            mixer_uninit(&mpctx->mixer);
+        mixer_uninit(&mpctx->mixer);
+        if (mpctx->ao)
             ao_uninit(mpctx->ao, mpctx->stop_play != AT_END_OF_FILE);
-        }
         mpctx->ao = NULL;
     }
 
-- 
1.7.12.3

